drivers/ukbus/pci/pci_bus.c

```diff
@@ -158,6 +167,273 @@ void _pci_register_driver(struct pci_driver *drv)
+		*size_out = ~size + 1;
+	else
+		*size_out = ~((uint32_t)size) + 1;
+
+	return 0;
+}
+
+/**
+ * Try to allocate the physical memory at @a paddr. This will allocate memory at
+ * another address if the region may be already in use.
+ * @param[inout] paddr the start of the physical memory range. Will be set the
+ *		       address to be used.
+ * @param pages the count of pages to allocate.
+ * @returns zero if the allocation at the specified address was successful, 1
+ *	    if space at a different address was allocated, and a negative errno
+ *	    in case the allocation was not successful.
+ */
+static int phys_alloc(__paddr_t *paddr, __sz pages)
+{
+#ifdef CONFIG_PAGING
+	struct uk_pagetable *pt;
+	int rc;
+
+	pt = ukplat_pt_get_active();
+
+	/* Allocate phys frames for the area. But don't allow that for NULL
+	 * addresses
+	 */
+	if (*paddr != 0)
+		rc = pt->fa->falloc(pt->fa, paddr, pages, 0);
+	else
+		rc = -ENOMEM;
+
+	if (rc == -ENOMEM) {
+		/* Range is already occupied. Move the PCI device */
+		*paddr = __PADDR_ANY;
+		rc = pt->fa->falloc(pt->fa, paddr, pages, 0);
+		if (rc)
+			return rc;
+		return 1;
+	} else if (rc && rc != -EFAULT) {
+		/* Some other error happened (and that error was not that the
+		 * range was outside the frame allocator range)
+		 */
+		return rc;
+	}
+
+	return 0;
+#else
+	void *a;
+
+	/* We have no idea what is used without a frame allocator, therefore
+	 * always allocate memory
+	 */
+	a = uk_malloc(uk_alloc_get_default(), pages * __PAGE_SIZE);
+	if (a == NULL)
+		return -ENOMEM;
+	*paddr = (__paddr_t)a;
+	return 1;
+#endif
+}
+
+static int physmem_free(__paddr_t paddr, __sz pages)
+{
+#ifdef CONFIG_PAGING
+	struct uk_pagetable *pt;
+
+	pt = ukplat_pt_get_active();
+	return pt->fa->ffree(pt->fa, paddr, pages);
+#else
+	(void)pages;
+	uk_free(uk_alloc_get_default(), (void *)paddr);
+	return 0;
+#endif
+}
+
+/* FIXME: We may have to track the maps, because for example msix_* needs an
+ * accessible MSI-X table.
+ */
+
+int pci_map_bar(struct pci_device *dev, __u8 idx, int attr,
+		struct pci_bar_memory *mem)
+{
+	__paddr_t bar_phys;
+	__vaddr_t bar_virt;
+	__sz bar_size;
+	__sz bar_pages;
+	int rc;
+#ifdef CONFIG_PAGING
+	struct uk_pagetable *pt;
+#endif
+
+	UK_ASSERT(mem);
+
+	/* Fetch BAR address and size of the memory region */
+	rc = pci_bar_phys_region(dev, idx, &bar_phys, &bar_size);
+	if (rc)
+		return rc;
+	bar_pages = DIV_ROUND_UP(bar_size, __PAGE_SIZE);
+
+	rc = phys_alloc(&bar_phys, bar_pages);
+	if (rc < 0)
+		return rc;
+	if (rc == 1) {
+		rc = pci_set_bar(dev, idx, bar_phys);
+		if (rc)
+			return rc;
+	}
+
+	/* Map base address memory */
+#ifdef CONFIG_PAGING
+	/* TODO: Allocate virtual address space range */
+	bar_virt = bar_phys;
+
+	pt = ukplat_pt_get_active();
+	uk_pr_debug("Mapping PCI device memory in virtual address space\n");
+	rc = ukplat_page_map(pt, bar_virt, bar_phys, bar_pages, attr, 0);
+	if (unlikely(rc)) {
+		pt->fa->ffree(pt->fa, bar_phys, bar_pages);
+		return ENOMEM;
```

Comment:
Shouldn't `rc` be returned here instead? I don't think `ENOMEM` is appropriate here.